### PR TITLE
Extracted the Maven Settings file

### PR DESCRIPTION
**Pull request by radtriste**

### CODE REVIEW

File: src/org/kie/jenkins/MavenSettingsConfig.groovy [hunk 1]
```diff
@@ -0,0 +1,119 @@
+package org.kie.jenkins
```
Comment (Ginxo) [inline src/org/kie/jenkins/MavenSettingsConfig.groovy:1]: This is probably the case for builder pattern, this is pretty easy on groovy. Just create a new `MavenSettingsConfigBuilder` file like

```
import groovy.transform.builder.Builder
import groovy.transform.builder.ExternalStrategy

// New builder class for the Message class.
@Builder(builderStrategy = ExternalStrategy, forClass = MavenSettingsConfig)
class MavenSettingsConfigBuilder {}
```
and then you will be able to...
```
def mavenSettingsConfig = new MavenSettingsConfigBuilder()  // Create new MavenSettingsConfigBuilder instance.
        .settingsXmlConfigFileId('whatever')
        .settingsXmlPath('whatever')
...
        .build()  // Return new MavenSettingsConfig instance. 
```

there is even a simpler option here (no need to create a new class for it)

```
import groovy.transform.builder.Builder
 
@Builder
class MavenSettingsConfig {
```
and then
`new MavenSettingsConfig.Builder().settingsXmlConfigFileId('whatever').build()`

but up to you

File: src/org/kie/jenkins/MavenSettingsConfig.groovy [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        if (this.disableSnapshotsInSettings) {
+            newCfg.withSnapshotsDisabledInSettings()
+        }
+        if (this.settingsXmlPath) {
+            newCfg.withSettingsXmlFile(this.settingsXmlPath)
+        }
+        return newCfg
+    }
+
+    String createSettingsFile() {
```
Comment (Ginxo) [inline src/org/kie/jenkins/MavenSettingsConfig.groovy:78]: By design we should keep this file as a simple POJO and move these `createSettingsFile` and `setRepositoryInSettings` methods to a new MavenSettinsService (example name) like

```
class MavenSettinsService {
  MavenSettingsConfig mavenSettingsConfig
  MavenSettinsService(MavenSettingsConfig mavenSettingsConfig) {
    this.mavenSettingsConfig = mavenSettingsConfig
  }
  
  String createSettingsFile() { ... }
  private void setRepositoryInSettings() { ... }

}
```

**Review by Ginxo — CHANGES_REQUESTED** [review]
please check

File: src/org/kie/jenkins/MavenSettingsConfig.groovy (hunk 2 above)
Comment (radtriste) [reply src/org/kie/jenkins/MavenSettingsConfig.groovy]: updated

**Comment by Ginxo** [comment]
I have the feeling methods liks `withSettingsXmlId` or `withSettingsXmlFile` are breaking the SRP and they should be moved to configuration file but I also understand the reason why for not complicating the things more. 
Anyway this is a good improvement that should be added in either way. So up to you to go for a builder pattern and so on or just merge the chenges.

**Comment by Ginxo** [comment]
Anyway the tests are failing
```
2021-12-16T15:54:09.4031422Z [INFO] Tests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.932 s - in MailerSpec
2021-12-16T15:54:09.7528693Z [INFO] 
2021-12-16T15:54:09.7535213Z [INFO] Results:
2021-12-16T15:54:09.7556692Z [INFO] 
2021-12-16T15:54:09.7557419Z [ERROR] Errors: 
2021-12-16T15:54:09.7561121Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] clone ok:138->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7564861Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] create with all set:28->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7568518Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] printSettings:169->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7572604Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withDependencyRepositoryInSettings:96->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7576867Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withMirrorDisabledForRepoInSettings:110->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7581084Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlFile and withSettingsXmlId:87->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7585018Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlFile:68->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7588853Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlId not existing:59->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7592977Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlId:50->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
2021-12-16T15:54:09.7630389Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSnapshotsDisabledInSettings:120->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
```

**Comment by radtriste** [comment]
> I have the feeling methods liks `withSettingsXmlId` or `withSettingsXmlFile` are breaking the SRP and they should be moved to configuration file but I also understand the reason why for not complicating the things more. Anyway this is a good improvement that should be added in either way. So up to you to go for a builder pattern and so on or just merge the chenges.

I added Builder class to MavenSettingsConfig

**Comment by radtriste** [comment]
> Anyway the tests are failing
> 
> ```
> 2021-12-16T15:54:09.4031422Z [INFO] Tests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.932 s - in MailerSpec
> 2021-12-16T15:54:09.7528693Z [INFO] 
> 2021-12-16T15:54:09.7535213Z [INFO] Results:
> 2021-12-16T15:54:09.7556692Z [INFO] 
> 2021-12-16T15:54:09.7557419Z [ERROR] Errors: 
> 2021-12-16T15:54:09.7561121Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] clone ok:138->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7564861Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] create with all set:28->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7568518Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] printSettings:169->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7572604Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withDependencyRepositoryInSettings:96->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7576867Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withMirrorDisabledForRepoInSettings:110->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7581084Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlFile and withSettingsXmlId:87->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7585018Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlFile:68->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7588853Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlId not existing:59->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7592977Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSettingsXmlId:50->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> 2021-12-16T15:54:09.7630389Z [ERROR]   MavenSettingsConfigSpec.[MavenSettingsConfig.groovy] withSnapshotsDisabledInSettings:120->JenkinsPipelineSpecification.addPipelineMocksToObjects_closure1$_closure13:533 » IllegalState
> ```

Let's see if corrected :)

**Comment by radtriste** [comment]
jenkins retest this

**Comment by radtriste** [comment]
@Ginxo not sure the `@Builder` and `@Autoclone` are available as CPS on Jenkins :(
See https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/blue/organizations/jenkins/KIE%2F7.59.x%2Fpullrequest%2Fjenkins-pipeline-shared-libraries-7.59.x.pr/detail/jenkins-pipeline-shared-libraries-7.59.x.pr/57/pipeline

**Comment by radtriste** [comment]
@Ginxo please have a review :)

File: src/org/kie/jenkins/MavenCommand.groovy [hunk 3]
```diff
@@ -41,50 +38,32 @@ def class MavenCommand {
...
-            }
         }
 
-        if(this.mavenOptions.size() > 0){
+        if (this.mavenOptions.size() > 0) {
             cmdBuilder.append(' ').append(this.mavenOptions.join(' '))
         }
         cmdBuilder.append(' ').append(goals)
-        if(this.profiles.size() > 0){
+        if (this.profiles.size() > 0) {
```
Comment (Ginxo) [inline src/org/kie/jenkins/MavenCommand.groovy:54]:
```suggestion
        if (this.profiles.size()) {
```

File: src/org/kie/jenkins/MavenCommand.groovy [hunk 4]
```diff
@@ -41,50 +38,32 @@ def class MavenCommand {
...
-                steps.sh "sed -i '/<pluginRepository>/,/<\\/pluginRepository>/ { /<snapshots>/,/<\\/snapshots>/ { s|<enabled>true</enabled>|<enabled>false</enabled>|; }}' ${settingsFile}"
-            }
-
-            if(this.printSettings){
-                steps.sh "cat ${settingsFile}"
-            }
         }
 
-        if(this.mavenOptions.size() > 0){
+        if (this.mavenOptions.size() > 0) {
```
Comment (Ginxo) [inline src/org/kie/jenkins/MavenCommand.groovy:50]:
```suggestion
        if (this.mavenOptions.size()) {
```

**Review by Ginxo — APPROVED** [review]
LGTM, just a couple of minor things. Well done!

File: src/org/kie/jenkins/MavenCommand.groovy (hunk 3 above)
Comment (radtriste) [reply src/org/kie/jenkins/MavenCommand.groovy]: updated

File: src/org/kie/jenkins/MavenCommand.groovy (hunk 4 above)
Comment (radtriste) [reply src/org/kie/jenkins/MavenCommand.groovy]: updated

File: src/org/kie/jenkins/MavenCommand.groovy [hunk 5]
```diff
@@ -93,39 +72,47 @@ def class MavenCommand {
...
+        this.settingsConfigBuilder = settingsConfigBuilder
+        return this
+    }
+
     /**
     * IF set, override `withSettingsXmlFile`
     **/
-    MavenCommand withSettingsXmlId(String settingsXmlId){
-        this.settingsXmlConfigFileId = settingsXmlId
+    MavenCommand withSettingsXmlId(String settingsXmlId) {
```
Comment (Ginxo) [inline src/org/kie/jenkins/MavenCommand.groovy:86]: The `with...` method returning `this` seems a clear Builder example, we should probably have an additional builder for MavenCommand like you did for MavenSettingsConfigBuilder.

> This is valid for the rest of `with...` methods

**Review by Ginxo — CHANGES_REQUESTED** [review]
it is obviously a great improvement, but I propose to add a new `MavenCommandBuilder` class and to move all the `with` methods from `MavenCommand` class to it.
I also miss BuilderSpec classes.

Great effort, congratulations.

File: src/org/kie/jenkins/MavenCommand.groovy (hunk 5 above)
Comment (radtriste) [reply src/org/kie/jenkins/MavenCommand.groovy:86]: @Ginxo doing so means we break a lot of pipelines using this code :S
I'd rather keep this interface on MavenCommand as it is not part of the PR.

But I could setup a MavenCommandBuilder and set as Deprecated the MavenCommand methods for now. wdyt ?

Comment (Ginxo) [reply src/org/kie/jenkins/MavenCommand.groovy:86]: I thought about that, and yes, you are right, we should keep MavenCommand API as it is, deprecate them, create a Builder class and move the code using the current code little by little.
I'm going to tag current repository status with `7.67.x`, this way we can use shared-library by using `@Library('jenkins-pipeline-shared-libraries@7.67.x')_`
@mbiarnes it is supposed the tag process also tag this repository as the rest of RHBA repositories

Comment (radtriste) [reply src/org/kie/jenkins/MavenCommand.groovy:86]: I can handle the Kogito pipelines refactoring with MavenCommandBuilder.

So gonna create 2 issues for my side:
- Move code to MavenCommandBuilder
- Refactor Kogito pipelines to use the MavenCommandBuilder

Is that ok for you ?

Comment (Ginxo) [reply src/org/kie/jenkins/MavenCommand.groovy:86]: perfect!! many thanks

Comment (radtriste) [reply src/org/kie/jenkins/MavenCommand.groovy:86]: created https://issues.redhat.com/browse/BXMSPROD-1701 and https://issues.redhat.com/browse/KOGITO-7192

**Comment by radtriste** [comment]
@Ginxo issues created. can we merge that one ? :)

**Review by Ginxo — APPROVED** [review]
LGTM

**Pull request merged**